group_counts et al now default to using the existing mask as the tabstops, unless explicitly given

DougBurke opened the pull request:
# Summary

When calling group_counts, group_width, ..., use the current mask as the tabStops argument by default (actually, the inverse of the mask), that is, when the tabStops argument is set to None (the default). This means that the grouping is only applied to the selected channels, and means that the filter range is less likely to change when filtered data is grouped.

This is a change in behaviour. To get the previous behaviour users need to set tabStops to an array of zeros (length being the number of channels): e.g. for Chandra ACIS imaging-mode data, group_counts(20, tabStops=[0] * 1024) or group_counts(20, tabStops=np.zeros(1024)).

Fix #417 #956 #1081

# Details

[stupid github interface. Fix #1081 ]

There are some minor clean up changes to allow 

- tabStops to be sent as a list rather than needing to be a ndarray, so that users can say `tabStops=[0] * 1024`
- if tabStops is sent a mask that matches the grouped size we allow this, automatically expanding this to the per-channel array in essentially the same way that the DataPHA get_mask routine does [*]; this is to make things easier for users as it has been an issue in my earlier experiments in getting tabStops to work (it's also the case that with follow-up changes it's unlikely to be useful for most users but it's not a large change)

[*] there are some possible issues with quality filtering here, but there's quality-filtering issues elsewhere too

The main change is so that if tabStops is None (the default) then we pick up the existing mask (inverted) and use that, so whereas before you would have to say

    group_counts(20, tabStops=~get_data().get_mask())

you can now say just

    group_counts(20)

This is a **breaking** change. If you want to get the old behavior you have to say either of

    group_counts(20, tabStops=[0] * 1024)
    group_counts(20, tabStops=np.zeros(1024))

changing 1024 to the number of channels in the dataset.

I spent some time trying to work out how to make this "backwards compatible", but I have come to the conclusion that it's actually better that we make this a breaking change - i.e. default to now only grouping within the noticed data range. However, this is something other people may think differently!

There is an argument to be said that we could have some "symbol" for `tabStops` which means "ignore the filter", so you don't have to say `tabStops=[0] * n` and can just say `tabStops="unfiltered"` (*as an example*).

---

# Example

In CIAO 4.15 we have

```
sherpa-4.15.0> load_pha("3c273.pi")
WARNING: systematic errors were not found in file '3c273.pi'
statistical errors were found in file '3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file 3c273.arf
read RMF file 3c273.rmf
WARNING: systematic errors were not found in file '3c273_bg.pi'
statistical errors were found in file '3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file 3c273_bg.pi

sherpa-4.15.0> ungroup()

sherpa-4.15.0> ignore(hi=0.5)
dataset 1: 0.00146:14.9504 -> 0.511:14.9504 Energy (keV)

sherpa-4.15.0> ignore(lo=7)
dataset 1: 0.511:14.9504 -> 0.511:6.9934 Energy (keV)

sherpa-4.15.0> get_filter()
'0.510999977589:6.993400096893'

sherpa-4.15.0> group_counts(20)

sherpa-4.15.0> get_filter()
'0.437999993563:13.461199760437'
```

So, after calling `group_counts` the filter has increased from 0.51:6.99 to 0.44:13.46 keV.

We can do the grouping only within the filtered range, but only by manually specifying the `tabStops` argument (in the correct manner):

```
sherpa-4.15.0> ungroup()
          ...: ignore(hi=0.5)
          ...: ignore(lo=7)
dataset 1: 0.438:13.4612 -> 0.511:13.4612 Energy (keV)
dataset 1: 0.511:13.4612 -> 0.511:6.9934 Energy (keV)

sherpa-4.15.0> get_filter()
'0.510999977589:6.993400096893'

sherpa-4.15.0> group_counts(20, tabStops=~get_data().get_mask())

sherpa-4.15.0> get_filter()
'0.510999977589:6.993400096893'
```

Note that the filter has not changed.

With this PR we get (you can see we get more information on the filter changes as we call the various group routines, due to #1637

```
In [1]: from sherpa.astro.ui import *
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'

In [2]: load_pha("3c273.pi")
WARNING: systematic errors were not found in file '3c273.pi'
statistical errors were found in file '3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file 3c273.arf
read RMF file 3c273.rmf
WARNING: systematic errors were not found in file '3c273_bg.pi'
statistical errors were found in file '3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file 3c273_bg.pi

In [3]: ungroup()
dataset 1: 0.00146:14.9504 Energy (keV) (unchanged)

In [4]: ignore(hi=0.5)
dataset 1: 0.00146:14.9504 -> 0.511:14.9504 Energy (keV)

In [5]: ignore(lo=7)
dataset 1: 0.511:14.9504 -> 0.511:6.9934 Energy (keV)

In [6]: group_counts(20)
dataset 1: 0.511:6.9934 Energy (keV) (unchanged)
```

So, the `group_counts` call (and all the other `group_xxx` ones) has automatically set `tabStops` for us.

If we want to behave like the old version we have to manually clear the `tabStops`: we need to know the number of channels - here 1024 - and say

```
In [7]: ungroup()
dataset 1: 0.511:6.9934 Energy (keV) (unchanged)

In [8]: notice()
dataset 1: 0.511:6.9934 -> 0.00146:14.9504 Energy (keV)

In [9]: ignore(lo=7)
dataset 1: 0.00146:14.9504 -> 0.00146:6.9934 Energy (keV)

In [10]: ignore(hi=0.5)
dataset 1: 0.00146:6.9934 -> 0.511:6.9934 Energy (keV)

In [11]: group_counts(20, tabStops=[0] * 1024)
dataset 1: 0.511:6.9934 -> 0.438:13.4612 Energy (keV)
```

DougBurke commented:
@anetasie and @hamogu - this is something I have wanted for quite a while - the `group_xxx` calls will not only group within the noticed data range. However, with the current approach it is a **breaking** change, in that users with 4.16 will get different results to 4.15. Fortunately the changes I added to report the filter when notice/ignore is called - #1562 in 4.15.0 - and group/... is called - #1637 in 4.15.1 - mean it is a bit more obvious to the user what is going on.

However, you may find that it is too invasive and we need another way to do this (i.e. do not do this by default) which is why it's still labelled a draft PR.

I do need to add some documentation, but your input is more-than-welcome here.

hamogu commented:
This is something that I've wanted for a while, too; probably since before I started contributing to sherpa development. In my opinion,  the old behavior is confusing and the new one much better.  Yes,  it's a breaking change,  but it will make it better.

I'm not making this as approved because I did not do a code review at this point.

DougBurke commented:
I think it's useful to just try it out as well, to make sure that it really does behave as you want @hamogu!

anetasie commented:
@DougBurke I have a perfect data set for the tabstops and will use it!

anetasie reviewed (approved):
Looks good.

DougBurke commented:
@anetasie - thanks for looking. One thing I wonder about is that if you want the "old" behaviur you have to explicitly set `tabStops` to an array of 0's - e.g. one of

```
group_counts(20, tabStops=[0] * 1024)
group_counts(20, tabStops=np.zeros(0, 1024))
```

This requires you to remember the number of channels and how to create an array of zeros. We could allow the user to set `tabStops` to a particular string - e.g.

```
group_counts(20, tabStops="nofilter")
```

Would this make sense to you? If so,what would be a good string to use (I don't want to use "none" as tabStops=None has a very-different behaviour).

anetasie commented:
@DougBurke this would be good for back compatibility. 

```
group_counts(20, tabStops="nofilter")
```

DougBurke commented:
For @anetasie - this "multi-modal" behaviour of keywords is generally frowned upon in the CS world, bit I think the UI layer has long-ago abandoned this for "ease of non-programmer use".

DougBurke commented:
@anetasie and @hamogu - I've updated this to

- add some documentation
- the ui layer lets you `group_counts(20, tabStops="nofilter")` to get the old behaviour (you can still say tabStops=[0] * nchan, for the appropriate nchan value, but this is to make it easier for users)
  - this is only in the UI layer; the DataPHA version does not support this  option (as it's an ease-of-use change we don't really want in the lower-level code)

Other than that, and all the typos I've added, this is ready for final review.

hamogu commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -472,20 +472,22 @@ def test_416_a():
 
     # We have a simplified mask
-    mask = [True, True]
+    mask = [False] * 3 + [True] * 2 + [False] * 4
```
This is exactly the mask defined in line 472. No need to put it in here again. Same is true for line 482/484 below.

DougBurke replied:
So, this is a hold over from when the mask was different. I'll clean up.

However I disagree with line 484 as that line is different (there's an extra bin).

DougBurke replied:
Actually, I mis-spole. This particular line *is* different, since the mask is now grouped. Hopefully the new version of the test will make it clearer.

hamogu commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -472,20 +472,22 @@ def test_416_a():
+    #
     pha.group_counts(3)
 
     # We have a simplified mask
```
We don't any longer...

hamogu commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -472,20 +472,22 @@ def test_416_a():
     assert pha.grouping == pytest.approx(grouping)
 
-    quality = [0, 0, 0, 0, 0, 2, 2, 2, 2, 2]
+    quality = [0, 0, 0, 0, 0, 2, 0, 0, 0, 0]
```
Wait, why has the quality changed? Why is the one in the middle quality 2? I don't see any quality array set anywhere, so I'm confused.

DougBurke replied:
So, the group call sets both the grouping and quality arrays. I've added text to the test to explain this, but we can view quality here as

  [3] * 0 + [0, 0, 2] + [0] * 4

The first 3 and last 4 channels are not grouped (since we've excluded them by the filter) and so are set to 0. The only channels we care about are the fourth to sixth channels, which have counts of [2, 1, 1]. The grouping groups them into 2 bins, so we have grouping of [1, -1, 1] to indicate this. The first group (of 2, 1) meets the group criteria (of >= 3) and so has a quality of 0. The last channel has a count of 1, which is < 3, and so has a quality of 3.

When we didn't include the tabstops we carried the 2's to the end of the array because these channels had counts of 0 so the group never hit the >= 3 criteria.
